activerecord/lib/active_record/associations/association.rb

```diff
@@ -377,10 +377,16 @@ def inversable?(record)
 
         def matches_foreign_key?(record)
           if foreign_key_for?(record)
-            record.read_attribute(reflection.foreign_key) == owner.id ||
-              (foreign_key_for?(owner) && owner.read_attribute(reflection.foreign_key) == record.id)
+            record.read_attribute(reflection.foreign_key) == owner.read_attribute(reflection.association_primary_key) ||
+              (foreign_key_for?(owner) && owner.read_attribute(reflection.foreign_key) == record.read_attribute(reflection.association_primary_key))
           else
-            owner.read_attribute(reflection.foreign_key) == record.id
+            record_primary_key = if reflection.polymorphic?
```

Comment:
On first glance, there doesnt seem to be a reason why this ought to be a reason why polymorphic associations cant be inversed, in fact the only trouble we get into is if we don't specify the primary_key, otherwise we wont fall into this else cases here:

https://github.com/rails/rails/blob/732a4748b361bfdadb042fc62b4c608b1a6eef47/activerecord/lib/active_record/reflection.rb#L941

and:

https://github.com/rails/rails/blob/732a4748b361bfdadb042fc62b4c608b1a6eef47/activerecord/lib/active_record/reflection.rb#L794

in fact it says here:
https://github.com/rails/rails/blob/732a4748b361bfdadb042fc62b4c608b1a6eef47/activerecord/lib/active_record/reflection.rb#L787

maybe...we should pass the class of the record in at the same time?

Reply:
I mean, some tests seem to say that they can be automatically inversed, and others not:
https://github.com/rails/rails/blob/732a4748b361bfdadb042fc62b4c608b1a6eef47/activerecord/test/cases/associations/inverse_associations_test.rb#L201-L215 Very confused.

Am running into this while doing an upgrade from Rails 5 T_T

Reply:
> On first glance, there doesnt seem to be a reason why this ought to be a reason why polymorphic associations cant be inversed

The reason I had to add the `reflection.polymorphic?` check because of https://github.com/rails/rails/blob/097f703d59a64da1164bf63ea020ff85e3b32004/activerecord/lib/active_record/reflection.rb#L446-L448

and fails this test `test_delete_polymorphic_has_many_with_destroy` with the following trace

```
ArgumentError: Polymorphic associations do not support computing the class.
    /Users/nickborromeo/nickborromeo/rails/activerecord/lib/active_record/reflection.rb:447:in `compute_class'
    /Users/nickborromeo/nickborromeo/rails/activerecord/lib/active_record/reflection.rb:406:in `klass'
    /Users/nickborromeo/nickborromeo/rails/activerecord/lib/active_record/reflection.rb:787:in `association_primary_key'
```